File: protocol/src/main/resources/js/ProtocolManager.js
const protocols = new Map();

const ProtocolManager = {};

ProtocolManager.getProtocol = function getProtocol(protocolId) {
    const protocol = protocols.get(protocolId);
    if (protocol === null) {
        throw new Error('[protocolId:' + protocolId + ']协议不存在');
    }
    return protocol;
};

ProtocolManager.write = function write(byteBuffer, packet) {
    const protocolId = packet.protocolId();
    byteBuffer.writeShort(protocolId);
    const protocol = ProtocolManager.getProtocol(protocolId);
    protocol.writeObject(byteBuffer, packet);
};

ProtocolManager.read = function read(byteBuffer) {
    const protocolId = byteBuffer.readShort();
    const protocol = ProtocolManager.getProtocol(protocolId);
    const packet = protocol.readObject(byteBuffer);
    return packet;
};

File: protocol/src/main/resources/lua/ProtocolManager.lua
local ByteBuffer = require("LuaProtocol.Buffer.ByteBuffer")

protocols = {}

ProtocolManager = {}

-- table扩展方法，后去set和map的大小
function table.setSize(set)
    local size = 0
    for _,_ in pairs(set) do
        size = size + 1
    end
    return size
end


function table.mapSize(map)
    local size = 0
    for _,_ in pairs(map) do
        size = size + 1
    end
    return size
end

function ProtocolManager.getProtocol(protocolId)
    local protocol = protocols[protocolId]
    if protocol == nil then
        error("[protocolId:" + protocolId + "]协议不存在")
    end
    return protocol
end

function ProtocolManager.write(buffer, packet)
    local protocolId = packet:protocolId()
    -- 写入协议号
    buffer:writeShort(protocolId)
    -- 写入包体
    ProtocolManager.getProtocol(protocolId):write(buffer, packet)
end

function ProtocolManager.read(buffer)
    local protocolId = buffer:readShort()
    return ProtocolManager.getProtocol(protocolId):read(buffer)
end

-- C#传进来的byte数组到lua里就会变成string
function readBytes(bytes)
    local buffer = ByteBuffer:new()
    buffer:writeBuffer(bytes)
    local packet = ProtocolManager.read(buffer)
    return packet
end

